### ISSUE TITLE
KL divergence expectations is returning “Infinity” as an observed value, which isn’t JSON-serializable

### ISSUE THREAD

**Issue by abegong** [fivetran/great_expectations#327]
(Replication code needed.)

**Comment by ccnobbli** [comment]
I'd be happy to take a look at this! Some context would be nice though

**Comment by abegong** [comment]
Actually, I'm not 100% sure this is a bug. As I wrote replication code, it worked okay.

Is there inconsistency in the way JSON parsers handle the token `Infinity`?

Here's replication code:
```
import json
import great_expectations as ge

#create a dataframe and partition
df = ge.dataset.PandasDataset({"x":[1,1,1,2,2,2,3,3,3]})
p = ge.dataset.util.categorical_partition_data(df.x)

#This expectation works just dandy
df.expect_column_kl_divergence_to_be_less_than("x", p, .05)

#This new dataframe has a value not seen before: 4
df2 = ge.dataset.PandasDataset({"x":[1,1,1,2,2,2,3,3,3,4]})

#Running the expectation works, but...
df2.expect_column_kl_divergence_to_be_less_than("x", p, .05)

#...now there's an "Infinity" in the JSON object
json.dumps(df2.expect_column_kl_divergence_to_be_less_than("x", p, .05))
```

`{"success": false, "result": {"observed_value": Infinity, "element_count": 10, "missing_count": 0, "missing_percent": 0.0}}`

**Comment by abegong** [comment]
More context: this was crashing in an application that uses GE for data validation. I'd need to do a bunch more digging to get to full replication...

**Comment by abegong** [comment]
Yep. It's a problem with PostgreSQL's JSON parser.

See here: https://github.com/encode/django-rest-framework/issues/5235

One fix that I'm seeing for this issue is to use `simplejson` with `ignore_nan`. That would lose a little bit of information (`infinity` vs `null`) but avoid crashiness. If we really wanted to, we could add something about Infinity in the `details` subobject.

**Comment by jpcampb2** [comment]
I think it might actually be that the python json parser is out of spec in allowing Infinity (we discussed this a while ago when we were talking about the input bounds acceptable for that test), but in either case, the question should be what ought we return in the case that the test is truly infinity, given that JSON technically doesn't support Infinity and python is just generous with us.

**Comment by abegong** [comment]
I vote we switch to `simplejson` and use the `ignore_nan` flag. That should take care of this.

https://simplejson.readthedocs.io/en/latest/

**Comment by jpcampb2** [comment]
I think that using simplejson + ignore_nan works since we are never differentiating between NaN, +Infinity and -Infinity in test output, and we engineered around it (using the tails concept) for input distributions. Another alternative that might be simpler, is to simply check in recursively convert to json serlializable for Infinity, and replace that with None. That way, we continue augmenting that helper function to be tuned to our specific case.

**Comment by anhollis** [comment]
I am trying to get a better understanding of this issue in order to work on potential solutions. Is "observed_value" holding infinity as a string (i.e. "infinity") or as something else? I am wondering if we can just set observed_value to "infinity" rather than python's inf, which I think is usually a float. Would this fix the problem, or is this just a restatement of the problem?

**Comment by abegong** [comment]
The issue is that the core JSON format specification doesn't have a concept of infinity.

Python's default json parser extends the core library and allows serialization and deserialization of Infinity values, but many other parsers don't allow this: they crash if they see a value of Infinity.

Rather than create confusion for other parsers by going outside the core JSON spec, we should just treat Infinity as None (which will serialize to `null`).

**Comment by anhollis** [comment]
I was just noticing that in addition to infinity showing up in the observed_value field of the output, we can also have infinity and -infinity show up in partition bins for continuous column input when result format is set to COMPLETE or SUMMARY. I assume that this will also be a problem for JSON to serialize.  Would changing partition endpoints to null when they are unbounded be an acceptable solution? That would seem really odd to me as a user if I saw that as result output for my partition object.

**Comment by abegong** [comment]
Good catch. Yes, that'll be a problem for most JSON serializers.

Can you include examples of what those objects look like? I'm wondering if there's a way to add string metadata to explain the infinity overflow problem...

**Comment by jpcampb2** [comment]
Yes--definitely a good catch. I'm worried about over-compensating for strict JSON a bit, but I do think we could do something simple to handle cases like this by adding an "inf_endpoints" parameter in the relevant partition objects. It might look something like this...

Current:
```
{
...
'details': {'observed_partition': {'bins': [-inf,
     -3.509954869848387,
     -0.7300588301209447,
     -0.050199927942877115,
     0.6135057959845841,
     2.8064786673806967,
     inf],
    'weights': [0.0, 0.25, 0.25, 0.25, 0.25, 0.0]},
   'expected_partition': {'bins': [-inf,
     -3.509954869848387,
     -0.7300588301209447,
     -0.050199927942877115,
     0.6135057959845841,
     2.8064786673806967,
     inf],
    'weights': [5e-06, 0.2499975, 0.2499975, 0.2499975, 0.2499975, 5e-06]}}
}
```

Proposed
```
'details': {'observed_partition': {
'inf_endpoints': 'both',
'bins': [
     -3.509954869848387,
     -0.7300588301209447,
     -0.050199927942877115,
     0.6135057959845841,
     2.8064786673806967],
    'weights': [0.0, 0.25, 0.25, 0.25, 0.25, 0.0]},
   'expected_partition': {'inf_endpoints': 'both',
'bins': [
     -3.509954869848387,
     -0.7300588301209447,
     -0.050199927942877115,
     0.6135057959845841,
     2.8064786673806967],
    'weights': [5e-06, 0.2499975, 0.2499975, 0.2499975, 0.2499975, 5e-06]}}}
```

where `'inf_endpoints'` values can be lower, upper, or both. It's more verbose to be sure, and would require extra work to deal with (including refactoring some of our internal code that currently pieces together the endpoints to allow such partition objects coming in to be valid).

**Comment by anhollis** [comment]
I would be up for trying to work on a solution to this, and I think it would be a good opportunity for me to get more familiar with the code base. I will come back to this forum with questions if I have them. Thanks!

**Comment by anhollis** [comment]
I actually just realized that the current code will always bookend the returned result object partitions with infinite endpoints. In other words, if the user gives the following partition as input: 

```
partition_object={"weights":[0.1,0.35,0.05,0.05,0.35,0.1],
        "bins":[-2,-1.5,-1,0,1,1.5, 2]} 
```

the result object will give the following output partitions:

```
'details': {'expected_partition': {'bins': [-inf,
     -2,
     -1.5,
     -1,
     0,
     1,
     1.5,
     2,
     inf],
    'weights': [0.0, 0.1, 0.35, 0.05, 0.05, 0.35, 0.1, 0.0]},
   'observed_partition': {'bins': [-inf, -2, -1.5, -1, 0, 1, 1.5, 2, inf],
    'weights': [0.0,
     0.0,
     0.30434782608695654,
     0.043478260869565216,
     0.21739130434782608,
     0.30434782608695654,
     0.13043478260869565,
     0.0]}}
```

With that in mind, I assume that we would want `inf_endpoints` to refer to the bins given by the user in the input `partition_object` rather than what is returned to the user in the result object. Is that right? If that is the case, then `inf_endpoints` could also be "neither", right? 

I was also thinking that since the output partitions will always have "tail bins" extending to positive and negative infinity, we could also output something like this:

```
'details': {'expected_partition': {'bins': [-2,
     -1.5,
     -1,
     0,
     1,
     1.5,
     2],
    'bin_weights': [ 0.1, 0.35, 0.05, 0.05, 0.35, 0.1],
    'tail_weights':[0.0, 0.0]},
   'observed_partition': {'bins': [ -2, -1.5, -1, 0, 1, 1.5, 2],
    'weights': [
     0.0,
     0.30434782608695654,
     0.043478260869565216,
     0.21739130434782608,
     0.30434782608695654,
     0.13043478260869565],
    'tail_weights':[0.0, 0.0]}}
```

**Comment by jpcampb2** [comment]
I *really* like your change: including `tail_weights` seems much clearer than the `inf_endpoints` nonsense. +1 for me for that.

Small clarification though: iirc, you won't *necessarily* get infinite endpoints from any partition...I think that adjustment only happens if you specify a nonzero `tail_weight_holdout` (please confirm).

**Comment by anhollis** [comment]
I ran the code on my computer and looked at the logic of the expectation function itself,  and it looks like the program does currently bookend `expected_partition` and `observed_partition` in `details` with inf and -inf regardless of whether or not a nonzero `tail_weight_holdout` is specified. 

Below is some replication code:

```
import great_expectations as ge

continuous_df=ge.dataset.PandasDataset({"x":[-1.5,-1.5,-1.35,-1.2,-1.19,-1.12,-1.04,-1,0,0,0,0.25,
                                             0.34,1,1.1,1.13,1.22,1.34,1.4,1.46,1.5,1.5,1.5]})

partition_object={"weights":[0.1,0.35,0.05,0.05,0.35,0.1],
        "bins":[-2,-1.5,-1,0,1,1.5, 2]}

continuous_df.expect_column_kl_divergence_to_be_less_than("x",partition_object, 0.05, result_format="COMPLETE")
```

I got the following output: 

```
{'result': {'details': {'expected_partition': {'bins': [-inf,
     -2,
     -1.5,
     -1,
     0,
     1,
     1.5,
     2,
     inf],
    'weights': [0.0, 0.1, 0.35, 0.05, 0.05, 0.35, 0.1, 0.0]},
   'observed_partition': {'bins': [-inf, -2, -1.5, -1, 0, 1, 1.5, 2, inf],
    'weights': [0.0,
     0.0,
     0.30434782608695654,
     0.043478260869565216,
     0.21739130434782608,
     0.30434782608695654,
     0.13043478260869565,
     0.0]}},
  'element_count': 23,
  'missing_count': 0,
  'missing_percent': 0.0,
  'observed_value': 0.2630026179072324},
 'success': False}
```

Note that I don't set tail_weight_holdout to a non-zero value. I can change this to match the behavior you were describing above, if you want.

**Comment by jpcampb2** [comment]
@anhollis We had a quick conversation about this on the call today:

1. Good catch: you're right that my memory was wrong about the addition of -inf and +inf on the bounds of bins for observed and expected partitions.

2. I really like your `tail_weights = [low_weight, high_weight]` option for dealing with the weights, replacing the alternate idea of inf_endpoints. Would you like to proceed down that path to create a PR?

**Comment by anhollis** [comment]
Yep, already working on a PR. Fingers crossed, I will put it in before next week.

**Comment by jcampbell** [comment]
This issue was addressed in #377 ; closing now.
